File: proto/tenants/Rue/Proto/Tenants/T3.hs
{-# LANGUAGE OverloadedStrings #-}
-- | T3: commit-confirmed firewall change (docs/ROADMAP.md section 8.3).
-- Permanent.
--
-- A region change to a host firewall with reach ssh(host); undo locus
-- :target; a backstop [unless_confirmed: 10m] installed and armed before the
-- change; a reachability probe; confirm(); commit() last. A Windows variant
-- with a firewall rule the tenant declares.
module Rue.Proto.Tenants.T3 (tenant, site, openMgmtPort, openMgmtPortWindows, pfAllow) where

import Rue.Proto.Model
import Rue.Proto.Tenants.Common

site :: Site
site =
  Site
    { siteHosts =
        [ HostRecord "fw-01" "freebsd" ["ssh"] True
        , HostRecord "fw-win-01" "windows" ["ssh"] True
        ]
    , siteTransports = ["ssh"]
    , siteAuthenticators = [Authenticator "netops" True]
    , siteMaxWait = Nothing
    , siteSchedulerPresent = ["fw-01", "fw-win-01"]
    }

pfAllow :: Op
pfAllow =
  (op "pf_allow" [anchored "file:/etc/pf.conf" "rue-mgmt"])
    { opUndoLocus = UndoTarget
    , opReach = ["ssh"]
    , opUndoOneLine = "strip the rue-mgmt anchor from /etc/pf.conf; pfctl reload"
    }

winfwAllow :: Op
winfwAllow =
  (op "winfw_allow" [entry Owned "winfw:rule:rue-mgmt"])
    { opUndoLocus = UndoTarget
    , opReach = ["ssh"]
    , opUndoOneLine = "remove the rue-mgmt firewall rule"
    }

confirmed :: Text' -> Op -> Plan
confirmed host change =
  (plan "open_mgmt_port" host [Step (step change) {stepArgs = ["port: 8443"]}, Observe "verify_reach" "reach", Confirm, Commit])
    { planBackstop = Just (Backstop [UnlessConfirmed (Duration 600)] (ArmBefore 1))
    }

type Text' = Host

openMgmtPort :: Plan
openMgmtPort = confirmed "fw-01" pfAllow

openMgmtPortWindows :: Plan
openMgmtPortWindows = confirmed "fw-win-01" winfwAllow

tenant :: Tenant
tenant =
  Tenant
    { tenantName = "t3"
    , tenantSite = site
    , tenantRequester = "netops_requester"
    , tenantCases = [Case "fw-01" openMgmtPort, Case "fw-win-01" openMgmtPortWindows]
    }

File: tenants/harness/src/tenants/t3.rs
//! T3: commit-confirmed firewall change (docs/ROADMAP.md section 8.3).
//! Permanent. A region change with reach ssh(host), undo locus :target, a
//! backstop [unless_confirmed: 10m] installed and armed before the change, a
//! reachability probe, confirm(), commit() last; and a Windows variant.

use rue_core::body::*;
use rue_core::model::*;

use super::common::*;

pub fn site() -> Site {
    Site {
        hosts: vec![
            host("fw-01", "freebsd", &["ssh"], true),
            host("fw-win-01", "windows", &["ssh"], true),
            // The same two firewalls with their artifacts in Python, and a
            // macOS pf host on the default sh: one plan, five artifacts.
            python_artifact(host("fw-02", "freebsd", &["ssh"], true)),
            python_artifact(host("fw-win-02", "windows", &["ssh"], true)),
            host("fw-mac-01", "macos", &["ssh"], true),
        ],
        transports: strings(&["ssh"]),
        authenticators: vec![authenticator("netops", true)],
        max_wait: None,
        scheduler_present: strings(&["fw-01", "fw-win-01", "fw-02", "fw-win-02", "fw-mac-01"]),
        secrets_deliver_to: vec![],
    }
}

pub fn pf_allow() -> Op {
    Op {
        undo_locus: UndoLocus::Target,
        reach: strings(&["ssh"]),
        do_: vec![
            region_set(
                anchored_ref("file:/etc/pf.conf", "rue-mgmt"),
                Value::Template(vec![
                    text("pass in proto tcp to port "),
                    interp(param("port")),
                ]),
            ),
            run_lit("pfctl -f /etc/pf.conf"),
        ],
        ..Op::new(
            "pf_allow",
            vec![FootprintEntry::anchored("file:/etc/pf.conf", "rue-mgmt")],
        )
    }
}

fn winfw_allow() -> Op {
    Op {
        undo_locus: UndoLocus::Target,
        reach: strings(&["ssh"]),
        do_: vec![run(vec![
            text("New-NetFirewallRule -Name rue-mgmt -Direction Inbound -Protocol TCP -LocalPort "),
            interp(param("port")),
            text(" -Action Allow"),
        ])],
        undo: computed(
            vec![run_lit("Remove-NetFirewallRule -Name rue-mgmt")],
            &["winfw:rule:rue-mgmt"],
        ),
        ..Op::new(
            "winfw_allow",
            vec![FootprintEntry::entry(Kind::Owned, "winfw:rule:rue-mgmt")],
        )
    }
}

fn confirmed(owner: &str, change: Op) -> Plan {
    Plan {
        backstop: Some(backstop(vec![Trigger::UnlessConfirmed(dur(600))], 1)),
        ..Plan::new(
            "open_mgmt_port",
            owner,
            vec![
                Item::Step(StepI {
                    args: strings(&["port: 8443"]),
                    ..StepI::new(change)
                }),
                Item::Observe {
                    probe: "verify_reach".into(),
                    alias: "reach".into(),
                },
                Item::Confirm,
                Item::Commit,
            ],
        )
    }
}

pub fn open_mgmt_port() -> Plan {
    confirmed("fw-01", pf_allow())
}

pub fn open_mgmt_port_windows() -> Plan {
    confirmed("fw-win-01", winfw_allow())
}

/// T3's body with the change step swapped, for the negatives derived from it.
pub fn body_with(change: Op) -> Vec<Item> {
    vec![
        with_args(change, &["port: 8443"]),
        Item::Observe {
            probe: "verify_reach".into(),
            alias: "reach".into(),
        },
        Item::Confirm,
        Item::Commit,
    ]
}

pub fn tenant() -> Tenant {
    Tenant {
        name: "t3".into(),
        site: site(),
        requester: "netops_requester".into(),
        cases: vec![
            case("fw-01", open_mgmt_port()),
            case("fw-win-01", open_mgmt_port_windows()),
            case(
                "fw-02",
                Plan {
                    owner: "fw-02".into(),
                    ..open_mgmt_port()
                },
            ),
            case(
                "fw-win-02",
                Plan {
                    owner: "fw-win-02".into(),
                    ..open_mgmt_port_windows()
                },
            ),
            case(
                "fw-mac-01",
                Plan {
                    owner: "fw-mac-01".into(),
                    ..open_mgmt_port()
                },
            ),
        ],
    }
}
